### DISCUSSION TITLE
Cluster mode

### DISCUSSION THREAD

**Discussion by uhlhosting** [caprover/caprover#1240] [Q&A] [answered]
Hi,

Recently I have attempted this scenario to get my first caprover cluster running.

I have cloned my machine entirely, increased the ip incrementaly +1 to avoid conflicts, and joined via docker swarm.
First, since they were two identical running caprovers, It asked me to remove existing swarm before joining a new one, when I joined I eneded up with this:

These are running now on Worker1
```
CONTAINER ID   IMAGE                                                  COMMAND                  CREATED         STATUS          PORTS     NAMES
5c56e330631c   registry.pac.uhl.cloud:996/captain/img-captain-you:0   "/docker-entrypoint.…"   5 minutes ago   Up 5 minutes    80/tcp    srv-captain--you.1.3j16wpq7uyl1m73a3ttzo84l2
a631078a8ada   caprover/netdata:v1.8.0                                "/run.sh"                3 days ago      Up 26 minutes             captain-netdata-container
root@cap:/captain/data# 
```

Now my question about the public IP of worker and the DNS, if my cluster its at *.pac.domain.tld what DNS I need for my Worker in place? How do I select when a deployment is made on worker? Or is this balanced automaticlally or?

**Comment by githubsaturn** [comment, marked as answer]
It's balanaced automatically. 

Having said that, you shouldn't have cloned the main machine. Worker node needs to be a clean installation of Linux + Docker.

**Reply by uhlhosting** [reply]
well when I went out of swarm it kinda become a clean machine, i have also runned prune on all leftovers in docker, and now this is unusable? what about the DNS part? Since they cant be on same ip, is it even needed for the second machine [worker] to have a public ip?

**Reply by githubsaturn** [reply]
>  i have also runned prune on all leftovers in docker, and now this is unusable? 

If you left swarm, and you pruned all containers, and you removed `/captain` directory, you're good to go.


> what about the DNS part?

You can point your URL to the ip address of the main node, the nginx container routes the request automatically if the app resides on the second node.


>  Since they cant be on same ip, is it even needed for the second machine [worker] to have a public ip?

No. Second node doesn't need a public IP. As long as it's on the same swarm network, it'll work just fine.

**Comment by uhlhosting** [comment]
Seems there is a bug somehow, since I removed a server from swarm, moved it to local IP, and rejoined the swarm: 

Current Cluster Nodes

```
Node ID:  
m3ptrcz5htvzzkymrkyo8f5x3
Type: worker
IP: 92.222.242.225/172.x.x.x
State: down
Status: active

RAM: 15.66 GB
OS: linux
CPU: 8 cores
Architecture: x86_64

Hostname: pac
Docker Version: 20.10.10
Node ID:  
sr4aeweo3mhqp7b384dxpe6x5
Type: worker
IP: 92.222.242.225/172.x.x.x
State: ready
Status: active

RAM: 15.66 GB
OS: linux
CPU: 8 cores
Architecture: x86_64

Hostname: pac
Docker Version: 20.10.10
Node ID:  
tnp2lukqtyx7nndcwsm2n36z4
Type: Leader (Main Node)
IP: 92.222.242.xxx
State: ready
Status: active

RAM: 15.66 GB
OS: linux
CPU: 8 cores
Architecture: x86_64

Hostname: pac
Docker Version: 20.10.10
```

How can I clean up the current clusters, and why isnt the local ip read properly? and instead it takes the first ip of my public IPV4 class, any ideas?

**Reply by uhlhosting** [reply]
Not to mention that the first two now are duplicates, or same server... How can I remove them and read?

**Reply by uhlhosting** [reply]
Seems after removing a server from swarm is not reflected in Master.

**Reply by githubsaturn** [reply]
What does Docker say?

Run this on your main node:
```
docker node ls
```

**Reply by uhlhosting** [reply]
I've done that already, it seems one was dead, or such, now I have only 2, yet the. ip is still bad reported.

**Reply by uhlhosting** [reply]
```
root@pac:/home/pac# docker node inspect sr4aeweo3mhqp7b384dxpe6x5
[
    {
        "ID": "sr4aeweo3mhqp7b384dxpe6x5",
        "Version": {
            "Index": 3432
        },
        "CreatedAt": "2021-11-01T22:47:57.093858594Z",
        "UpdatedAt": "2021-11-01T23:03:06.68155972Z",
        "Spec": {
            "Labels": {},
            "Role": "worker",
            "Availability": "active"
        },
        "Description": {
            "Hostname": "pac",
            "Platform": {
                "Architecture": "x86_64",
                "OS": "linux"
            },
            "Resources": {
                "NanoCPUs": 8000000000,
                "MemoryBytes": 16818495488
            },
            "Engine": {
                "EngineVersion": "20.10.10",
                "Plugins": [
                    {
                        "Type": "Log",
                        "Name": "awslogs"
                    },
                    {
                        "Type": "Log",
                        "Name": "fluentd"
                    },
                    {
                        "Type": "Log",
                        "Name": "gcplogs"
                    },
                    {
                        "Type": "Log",
                        "Name": "gelf"
                    },
                    {
                        "Type": "Log",
                        "Name": "journald"
                    },
                    {
                        "Type": "Log",
                        "Name": "json-file"
                    },
                    {
                        "Type": "Log",
                        "Name": "local"
                    },
                    {
                        "Type": "Log",
                        "Name": "logentries"
                    },
                    {
                        "Type": "Log",
                        "Name": "splunk"
                    },
                    {
                        "Type": "Log",
                        "Name": "syslog"
                    },
                    {
                        "Type": "Network",
                        "Name": "bridge"
                    },
                    {
                        "Type": "Network",
                        "Name": "host"
                    },
                    {
                        "Type": "Network",
                        "Name": "ipvlan"
                    },
                    {
                        "Type": "Network",
                        "Name": "macvlan"
                    },
                    {
                        "Type": "Network",
                        "Name": "null"
                    },
                    {
                        "Type": "Network",
                        "Name": "overlay"
                    },
                    {
                        "Type": "Volume",
                        "Name": "local"
                    }
                ]
            },
            "TLSInfo": {
                "TrustRoot": "-----BEGIN CERTIFICATE-----\/\\\+/+\/\/\+\++q\n-----END CERTIFICATE-----\n",
                "CertIssuerSubject": "",
                "CertIssuerPublicKey": "++=="
            }
        },
        "Status": {
            "State": "ready",
            "Addr": "92.222.242.225"
        }
    }
]
root@pac:/home/pac# 

```

**Reply by githubsaturn** [reply]
So CapRover itself doesn't do anything specific with node management. It's all pure docker. Specially this:
> Seems after removing a server from swarm is not reflected in Master.

When you go to Cluster tab, CapRover simply asks Docker engine to return the nodes, and then CapRover displays it on the web dashboard.

If you have reproducible steps for this issue, I suggest you post it on Docker's issue tracker https://github.com/moby/moby/issues/
